File: backend/app/schemas/scheduling.py
from pydantic import BaseModel, Field, field_validator
from typing import List, Optional, Dict, Any
from datetime import datetime
from enum import Enum


class ValidationResultEnum(str, Enum):
    ALLOWED = "allowed"
    BLOCKED = "blocked"
    WARNED = "warned"


class ActionEnum(str, Enum):
    ALLOW = "allow"
    BLOCK = "block"
    WARN = "warn"
    SUGGEST_ALTERNATIVE = "suggest_alternative"


class ValidationResult(BaseModel):
    is_valid: bool
    validation_result: ValidationResultEnum
    warnings: List[str] = []
    block_reasons: List[str] = []
    suggestions: List["SuggestionSlot"] = []
    rule_evaluations: List["RuleEvaluationResult"] = []


class SuggestionSlot(BaseModel):
    start_time: datetime
    end_time: datetime
    score: float = Field(..., ge=0.0, le=1.0)
    reason: str
    calendar_day_id: Optional[str] = None


class RuleEvaluationResult(BaseModel):
    rule_id: str
    rule_name: str
    action: ActionEnum
    triggered: bool
    message: Optional[str] = None
    severity: str = "info"


class SchedulingRuleCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = None
    conditions: List[Dict[str, Any]]
    action: ActionEnum
    alert_message: Optional[str] = None
    priority_order: int = Field(default=100, ge=1, le=1000)
    is_active: bool = True


class SchedulingRuleResponse(SchedulingRuleCreate):
    rule_id: str
    created_at: datetime
    updated_at: datetime


class ValidationRequest(BaseModel):
    task_id: str
    proposed_start_time: str  # Local time string in format YYYY-MM-DDTHH:MM:SS
    proposed_end_time: str    # Local time string in format YYYY-MM-DDTHH:MM:SS

    @field_validator('proposed_start_time', 'proposed_end_time')
    @classmethod
    def parse_local_datetime(cls, v: str) -> datetime:
        """Parse local time string as datetime object (treating as local time, not UTC)"""
        try:
            # Parse the local time string as a naive datetime (no timezone info)
            return datetime.fromisoformat(v)
        except ValueError as e:
            raise ValueError(f"Invalid datetime format. Expected YYYY-MM-DDTHH:MM:SS, got: {v}") from e


class SuggestionRequest(BaseModel):
    task_id: str
    start_date: str  # YYYY-MM-DD
    end_date: str    # YYYY-MM-DD

class AutoScheduleRequest(BaseModel):
    task_ids: List[str]
    start_date: str  # YYYY-MM-DD
    end_date: str    # YYYY-MM-DD 

File: frontend/src/api/scheduling.ts
import axios from 'axios'

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000/api/v1'

const api = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
  },
})

api.interceptors.request.use((config) => {
  const token = localStorage.getItem('access_token')
  if (token) {
    config.headers = config.headers || {}
    config.headers['Authorization'] = `Bearer ${token}`
  }
  return config
})

// TypeScript interfaces matching our backend schemas
export interface ValidationResult {
  is_valid: boolean
  validation_result: string
  warnings: string[]
  block_reasons: string[]
  suggestions: SuggestionSlot[]
  rule_evaluations: RuleEvaluationResult[]
}

export interface SuggestionSlot {
  start_time: string
  end_time: string
  score: number
  reason: string
  calendar_day_id?: string
}

export interface RuleEvaluationResult {
  rule_id: string
  rule_name: string
  action: 'allow' | 'block' | 'warn' | 'suggest_alternative'
  triggered: boolean
  message?: string
  severity: string
}

export type ConditionOperator =
  | 'equals'
  | 'not_equals'
  | 'greater_than'
  | 'less_than'
  | 'in'
  | 'not_in'

export interface RuleCondition {
  source: 'task_property' | 'calendar_day' | 'time_slot'
  field: string
  operator: ConditionOperator
  value: string | number | boolean | string[]
}

export interface AutoScheduleResult {
  scheduled_tasks: Array<{ task_id: string; title: string; scheduled_time: string; score: number }>
  failed_tasks: Array<{ task_id: string; title: string; reason: string; errors?: string[] }>
}

export interface SchedulingRule {
  rule_id: string
  name: string
  description?: string
  conditions: RuleCondition[]
  action: 'allow' | 'block' | 'warn' | 'suggest_alternative'
  alert_message?: string
  priority_order: number
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface ValidationRequest {
  task_id: string
  proposed_start_time: string
  proposed_end_time: string
}

export interface SuggestionRequest {
  task_id: string
  start_date: string
  end_date: string
}

export interface AutoScheduleRequest {
  task_ids: string[]
  start_date: string
  end_date: string
}

export interface RuleBuilderConfig {
  categories: Array<{ value: string; label: string }>
  work_environments: Array<{ value: string; label: string }>
  focus_levels: Array<{ value: string; label: string }>
  time_slot_focus_levels: Array<{ value: string; label: string }>
  task_priorities: Array<{ value: string; label: string }>
  boolean_options: Array<{ value: boolean; label: string }>
  task_properties: Array<{ value: string; label: string }>
  calendar_day_properties: Array<{ value: string; label: string }>
  time_slot_properties: Array<{ value: string; label: string }>
  operators: Record<string, Array<{ value: string; label: string }>>
}

// API functions
export async function validatePlacement(
  taskId: string,
  startTime: string,
  endTime: string
): Promise<ValidationResult> {
  const response = await api.post('/scheduling/validate', {
    task_id: taskId,
    proposed_start_time: startTime,
    proposed_end_time: endTime,
  })
  return response.data
}

export async function suggestSlots(
  taskId: string,
  startDate: string,
  endDate: string
): Promise<SuggestionSlot[]> {
  const response = await api.post('/scheduling/suggest', {
    task_id: taskId,
    start_date: startDate,
    end_date: endDate,
  })
  return response.data
}

export async function autoScheduleTasks(
  taskIds: string[],
  startDate: string,
  endDate: string
): Promise<AutoScheduleResult> {
  const response = await api.post('/scheduling/auto-schedule', {
    task_ids: taskIds,
    start_date: startDate,
    end_date: endDate,
  })
  return response.data
}

export async function getRuleBuilderConfig(): Promise<RuleBuilderConfig> {
  const response = await api.get('/scheduling/rule-builder-config')
  return response.data
}

export async function getSchedulingRules(): Promise<SchedulingRule[]> {
  const response = await api.get('/scheduling/rules')
  return response.data
}

export async function createSchedulingRule(
  rule: Omit<SchedulingRule, 'rule_id' | 'created_at' | 'updated_at'>
): Promise<SchedulingRule> {
  const response = await api.post('/scheduling/rules', rule)
  return response.data
}

export async function updateSchedulingRule(
  ruleId: string,
  rule: Partial<SchedulingRule>
): Promise<SchedulingRule> {
  const response = await api.put(`/scheduling/rules/${ruleId}`, rule)
  return response.data
}

export async function deleteSchedulingRule(ruleId: string): Promise<void> {
  await api.delete(`/scheduling/rules/${ruleId}`)
}
